docker/Makefile

```diff
@@ -0,0 +1,213 @@
+DOCKER_JUPYTER_COMMAND += -m $(JUPYTER_MODE)
+
+stop_jupyter :
+	- docker rm -f jupyter_container
+
+run_jupyter : build_contracts stop_jupyter
+	docker run $(DOCKER_RUN_ARGS) -it --name jupyter_container pdo_contracts:${CONTRACTS_VERSION} \
+		$(DOCKER_JUPYTER_COMMAND)
+
+# -----------------------------------------------------------------
+# We need a repository with the source for the branch we are going
+# to build. In theory this could just be a copy of the local source
+# tree but it might contain other artifacts. So instead we clone
+# the contracts with just a single branch in it. The storage and
+# performance requirements are relatively low.
+# -----------------------------------------------------------------
+contracts :
+	git clone --single-branch --branch $(CONTRACTS_BRANCH) --recurse-submodules $(CONTRACTS_REPO) contracts
+
+clean_contracts :
+	rm -rf contracts
+
+# -----------------------------------------------------------------
+# Testing uses docker-compose to build a set of images that will be configured
+# to run with localhost as the interface. in order to get a clean shutdown we
+# don't use a docker-compose specification for the base, ccf_base and services_base
+# containers. since we are rebuilding the contracts, we force recompilation of these
+# images from the rebuilt contracts. the corresponding images are built using docker
+# build instead of being included in the docker-compose.
+# -----------------------------------------------------------------
+
+# TEST_ARGS provides additional parameters for the compose
+# operations for running the test environment; generally
+# this would set in a make.loc
+TEST_COMPOSE_ARGS += -f test.yaml
+
+build_test : contracts models build_contracts versions
+
+test : clean_config clean_contracts build_test
+	- PDO_VERSION=$(PDO_VERSION) CONTRACTS_VERSION=$(CONTRACTS_VERSION) \
+	 	docker-compose $(TEST_COMPOSE_ARGS) up --abort-on-container-exit
+	PDO_VERSION=$(PDO_VERSION) CONTRACTS_VERSION=$(CONTRACTS_VERSION) \
+	 	docker-compose $(TEST_COMPOSE_ARGS) down
+
+JUPYTER_COMPOSE_ARGS += -f test_jupyter.yaml
+
+test_jupyter : clean_config clean_contracts build_test
+	- PDO_VERSION=$(PDO_VERSION) CONTRACTS_VERSION=$(CONTRACTS_VERSION) \
+		docker-compose $(JUPYTER_COMPOSE_ARGS) up --abort-on-container-exit
+	PDO_VERSION=$(PDO_VERSION) CONTRACTS_VERSION=$(CONTRACTS_VERSION) \
+		docker-compose $(JUPYTER_COMPOSE_ARGS) down
+
+# -----------------------------------------------------------------
+# Cleaning is a bit interesting because the containers don't go away
+# unless they are told to very nicely. Until they go away they hold onto
+# referenced images. Clean make sure that everything is cleaned up
+# across the board.
+# -----------------------------------------------------------------
+_IMAGES_=$(shell docker images -a --filter=dangling=true -q)
+_CONTAINERS_=$(shell docker ps --filter=status=exited --filter=status=created -q)
```

Comment:
Hmm, that's a bit a brutal indiscriminate killing/removing of all containers?!  E.g., i share now icx server with prakash and others and that would kill everyones containers/images?

Reply:
you do realize that this is your code, right? :)

seriously, this came from the original PDO docker makefile (or something like that). it is also in the current pdo docker makefile (for better or worse).

and yes i know its rather... indiscriminate. i'm ok with that.  see my note above about the point of this directory is to provide a complete, clean and (to the extent possible for the contracts tests) automate test environment. cleaning all containers is pretty "thorough" but it does make sure that the environment is clean. 

again... i'm open to suggestions (like maybe adding -rm to the start), but i don't think "make it work for two people on the same vm" is part of the requirements for this (i don't think it would work anyway since we fix the name of the containers).

Reply:
> you do realize that this is your code, right? :)

I don't think this is true (as in this was in pdo code before the refactor)

> seriously, this came from the original PDO docker makefile (or something like that). it is also in the current pdo docker makefile (for better or worse).

Ah, now i notice, indeed.  Looking a bit closer i realize though it's a bit less discriminating than i originally thought. it deletes any container but only dangling/exited ones, so kindof should be safe.

> and yes i know its rather... indiscriminate. i'm ok with that. see my note above about the point of this directory is to provide a complete, clean and (to the extent possible for the contracts tests) automate test environment. cleaning all containers is pretty "thorough" but it does make sure that the environment is clean.
> 
> again... i'm open to suggestions (like maybe adding -rm to the start), but i don't think "make it work for two people on the same vm" is part of the requirements for this (i don't think it would work anyway since we fix the name of the containers).

Well, that would be quite easy to addess as raised as a question in my docker improvement PR (PDO PR 242): just adopt what we already suggest in the docu and integrate the user name into the container name? (see that PR's description for a few more thoughts on that).   Whether that's worth the effort of course is a different question ...

Reply:
i do not want to incorporate user names into the test process. its really easy to write your own compose files with the names in them for development.

Reply:
cleaned up the clean and stop targets. should be easier to understand if nothing else.